### ISSUE TITLE
DefaultIfEmpty exception in EntityFrameworkCore

### ISSUE THREAD

**Issue by fengwusheng** [dotnet/efcore#14902]
The same code using MS Sql Server 2008:
```C#
string value = context.Book.Where(item => false).DefaultIfEmpty().Select(item => "520").FirstOrDefault();
```

With Microsoft.EntityFrameworkCore.2.2.2 get exception "Object reference not set to an instance of an object." and tsql translated is:
```sql
SELECT TOP(1) 1
FROM (
　SELECT NULL AS [empty]
) AS [empty]
LEFT JOIN (
　SELECT [item].*
　FROM [Book] AS [item]
　WHERE 0 = 1
) AS [t] ON 1 = 1
```

With EF5 get the right result, value is "520" and tsql translated is:
```sql
SELECT TOP (1) 
N'520' AS [C1]
FROM   ( SELECT 1 AS X ) AS [SingleRowTable1]
LEFT OUTER JOIN  (SELECT [SingleRowTable2].[X] AS [X]
　FROM  ( SELECT 1 AS X ) AS [SingleRowTable2]
　WHERE 1 = 0 ) AS [Filter1] ON 1 = 1
```

**Comment by Zenexer** [comment]
As of EF Core 3.0.0-preview6, this is now results in a NotImplementedException: https://github.com/aspnet/EntityFrameworkCore/blob/3d44d35b1c5eec5f5bc5253cca56c57beee98f0d/src/EFCore.Relational/Query/Pipeline/RelationalQueryableMethodTranslatingExpressionVisitor.cs#L232

**Comment by Zenexer** [comment]
Temporary workarounds to achieve MinAsync and MaxAsync; replace:

``` csharp
var min = await db.Items
    .Where(x => x.Condition)
    .DefaulfEmpty()
    .Select(x => x.Value)
    .MinAsync(cancellationToken);

var max= await db.Items
    .Where(x => x.Condition)
    .DefaulfEmpty()
    .Select(x => x.Value)
    .MaxAsync(cancellationToken);
```

with:

``` csharp
var min = await db.Items
    .Where(x => x.Condition)
    .OrderBy(x => x.Value)
    .Select(x => x.Value)
    .FirstOrDefaultAsync(cancellationToken);

var max = await db.Items
    .Where(x => x.Condition)
    .OrderByDescending(x => x.Value)
    .Select(x => x.Value)
    .FirstOrDefaultAsync(cancellationToken);
```

**Comment by smitpatel** [comment]
This issue is fixed in 3.1
Generated SQL
```SQL
      SELECT TOP(1) N'520'
      FROM (
          SELECT NULL AS [empty]
      ) AS [empty]
      LEFT JOIN (
          SELECT [b].[Id]
          FROM [Book] AS [b]
          WHERE CAST(0 AS bit) = CAST(1 AS bit)
      ) AS [t] ON 1 = 1
```
